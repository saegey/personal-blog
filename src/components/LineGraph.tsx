import { Line } from '@nivo/line'
import { ResponsiveWrapper } from '@nivo/core'
import { useThemeUI } from 'theme-ui'

import { formatSeconds, formatTime } from '../lib/formatters'
import { LineGraphProps } from '../common/types'
import { themeTemplate } from '../lib/graphHelper'

export const LineGraph = ({
  data,
  unit,
  yScaleMin,
  yScaleMax,
  areaBaselineValue,
  axisBottomTickValues,
  axisLeftTickValues,
  curve,
  enableArea = true,
  lineWidth = 1,
  colors,
  markers,
  xScaleType,
}: LineGraphProps) => {
  const { theme } = useThemeUI()
  const themeData = themeTemplate(theme)

  return (
    <Line
      yScale={{
        type: 'linear',
        min: yScaleMin,
        max: yScaleMax,
        stacked: false,
        reverse: false,
      }}
      xScale={{
        type: xScaleType ? xScaleType : 'linear',
        min: 'auto',
        max: 'auto',
      }}
      margin={{
        top: 20,
        right: 10,
        bottom: 25,
        left: 40,
      }}
      curve={curve ? curve : 'natural'}
      pointSize={0}
      useMesh
      enableArea={enableArea}
      enableGridX={false}
      lineWidth={lineWidth}
      areaBaselineValue={areaBaselineValue}
      areaOpacity={1.0}
      colors={colors ? colors : [theme.colors?.graphFill as string]}
      axisBottom={{
        format: formatSeconds,
        tickSize: 0,
        tickPadding: 5,
        tickRotation: 0,
        // legend: "Time",
        legendOffset: 30,
        legendPosition: 'middle',
        tickValues: axisBottomTickValues,
      }}
      enableSlices="x"
      sliceTooltip={({ slice }) => (
        <div
          style={{
            backgroundColor: theme.colors?.text,
            padding: '9px 12px',
            fontFamily: theme.fonts?.body,
            letterSpacing: '.4px',
            borderRadius: '4px',
          }}
        >
          <div
            key={slice.points[0].id}
            style={{
              color: theme.colors?.background,
              padding: '3px 0',
            }}
          >
            <div style={{ fontWeight: 300 }}>
              {formatTime(Number(slice.points[0].data.x))}
            </div>
            <div style={{ fontWeight: 600 }}>
              {slice.points[0].data.y.toLocaleString()} {unit}
            </div>
          </div>
        </div>
      )}
      axisLeft={{
        format: val => `${val}`,
        tickSize: 0,
        tickPadding: 5,
        tickRotation: 0,
        // legend: "Elevation",
        legendOffset: 0,
        legendPosition: 'middle',
        tickValues: axisLeftTickValues,
      }}
      markers={markers}
      theme={themeData}
      layers={[
        'grid',
        'markers',
        'axes',
        'areas',
        'crosshair',
        'lines',
        'slices',
        'mesh',
        'legends',
      ]}
      data={data}
    />
  )
}

const LineGraphResponsive = (props: LineGraphProps) => (
  <ResponsiveWrapper>
    {({ width, height }) => <Line width={width} height={height} {...props} />}
  </ResponsiveWrapper>
)

export default LineGraphResponsive
