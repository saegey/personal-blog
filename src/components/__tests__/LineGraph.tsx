/**
 * @jest-environment jsdom
 */
import renderer from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react'

import { ThemeProvider } from 'theme-ui'

import LineGraph from '../LineGraph'

import { theme } from './__test-utils__'

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}))

describe('LineGraph', () => {
  test.skip('renders', () => {
    const json = renderer
      .create(
        <ThemeProvider theme={theme}>
          <LineGraph
            data={[
              {
                id: 'testdata',
                data: [
                  { x: 1, y: 10 },
                  { x: 2, y: 5 },
                  { x: 3, y: 15 },
                ],
              },
            ]}
            yScaleMin={0}
            yScaleMax={20}
            colors={['black']}
            areaBaselineValue={0}
            axisBottomTickValues={[0, 1, 2, 3]}
            axisLeftTickValues={[0, 5, 10, 15, 20]}
            enableArea={false}
            unit={'watts'}
            lineWidth={2}
          />
        </ThemeProvider>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  // it('renders tooltip', async () => {
  //   render(
  //     <ThemeProvider theme={theme}>
  //       <LineGraph
  //         data={[
  //           {
  //             id: 'testdata',
  //             data: [
  //               { x: 1, y: 10 },
  //               { x: 2, y: 5 },
  //               { x: 3, y: 15 },
  //             ],
  //           },
  //         ]}
  //         yScaleMin={0}
  //         yScaleMax={20}
  //         areaBaselineValue={0}
  //         axisBottomTickValues={[0, 1, 2, 3]}
  //         axisLeftTickValues={[0, 5, 10, 15, 20]}
  //         enableArea={false}
  //         unit={'watts'}
  //         lineWidth={2}
  //       />
  //     </ThemeProvider>
  //   )
  // const blah = await screen.queryAllByRole('test')
  // console.log(blah)
  // screen.debug()
  // })
})
