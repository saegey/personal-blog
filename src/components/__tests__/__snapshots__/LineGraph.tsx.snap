// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineGraph renders 1`] = `
<div
  style={
    {
      "position": "relative",
    }
  }
>
  <svg
    role="img"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      fill="transparent"
    />
    <g
      transform="translate(40,20)"
    >
      <g>
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
        <line
          opacity={1}
          stroke="#dddddd"
          strokeWidth={1}
          x1={0}
          x2={NaN}
          y1={NaN}
          y2={NaN}
        />
      </g>
      <g
        transform="translate(0,NaN)"
      >
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(NaN,0)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="text-before-edge"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="middle"
            transform="translate(0,5) rotate(0)"
          >
            0s
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(NaN,0)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="text-before-edge"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="middle"
            transform="translate(0,5) rotate(0)"
          >
            1s
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(NaN,0)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="text-before-edge"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="middle"
            transform="translate(0,5) rotate(0)"
          >
            2s
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(NaN,0)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="text-before-edge"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="middle"
            transform="translate(0,5) rotate(0)"
          >
            3s
          </text>
        </g>
        <line
          style={
            {
              "stroke": "transparent",
              "strokeWidth": 1,
            }
          }
          x1={0}
          x2={NaN}
          y1={0}
          y2={0}
        />
      </g>
      <g
        transform="translate(0,0)"
      >
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(0,NaN)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="end"
            transform="translate(-5,0) rotate(0)"
          >
            0
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(0,NaN)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="end"
            transform="translate(-5,0) rotate(0)"
          >
            5
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(0,NaN)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="end"
            transform="translate(-5,0) rotate(0)"
          >
            10
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(0,NaN)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="end"
            transform="translate(-5,0) rotate(0)"
          >
            15
          </text>
        </g>
        <g
          style={
            {
              "opacity": 1,
            }
          }
          transform="translate(0,NaN)"
        >
          <line
            style={
              {
                "stroke": "#777777",
                "strokeWidth": 1,
              }
            }
            x1={0}
            x2={-0}
            y1={0}
            y2={0}
          />
          <text
            dominantBaseline="central"
            style={
              {
                "fill": "#333333",
                "fontFamily": "bodyfont",
                "fontSize": 14,
              }
            }
            textAnchor="end"
            transform="translate(-5,0) rotate(0)"
          >
            20
          </text>
        </g>
        <line
          style={
            {
              "stroke": "transparent",
              "strokeWidth": 1,
            }
          }
          x1={0}
          x2={0}
          y1={0}
          y2={NaN}
        />
      </g>
      <path
        d="MNaN,NaNCNaN,NaN,NaN,NaN,NaN,NaNCNaN,NaN,NaN,NaN,NaN,NaN"
        fill="none"
        strokeWidth={2}
      />
      <rect
        fill="red"
        fillOpacity={0}
        height={NaN}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        stroke="red"
        strokeOpacity={0.75}
        strokeWidth={0}
        width={NaN}
        x={NaN}
        y={0}
      />
    </g>
  </svg>
</div>
`;
